src/lib/optimizer/strategy/predicate_placement_rule.cpp

```diff
@@ -88,10 +95,118 @@ void PredicatePlacementRule::_push_down_traversal(const std::shared_ptr<Abstract
+                    left_conjunction.emplace_back(expression_in_conjunction);
+                  }
+                  if (evaluable_on_right_side && !evaluable_on_left_side && !aborted_right_side) {
+                    right_conjunction.emplace_back(expression_in_conjunction);
+                  }
+                }
+
+                if (!left_conjunction.empty()) {
+                  // If we have found one or more predicates for the left side, connect them using AND and add them to
+                  // the disjunction that will be pushed to the left side.
+                  left_disjunction.emplace_back(inflate_logical_expressions(left_conjunction, LogicalOperator::And));
+                } else {
+                  // If, within the current expression_in_disjunction, we have not found a matching predicate for the
+                  // left side, all tuples for the left side qualify and it makes no sense to create a filter.
+                  aborted_left_side = true;
+                  left_disjunction.clear();
+                }
+                if (!right_conjunction.empty()) {
+                  right_disjunction.emplace_back(inflate_logical_expressions(right_conjunction, LogicalOperator::And));
+                } else {
+                  aborted_right_side = true;
+                  right_disjunction.clear();
+                }
+              }
+
+              const auto add_disjunction_if_beneficial =
+                  [&](const auto& disjunction, const auto& disjunction_input_node, auto& predicate_nodes) {
+                    if (disjunction.empty()) return;
+
+                    const auto expression = inflate_logical_expressions(disjunction, LogicalOperator::Or);
+                    const auto predicate_node = PredicateNode::make(expression, disjunction_input_node);
+
+                    // Determine the selectivity of the predicate if executed on disjunction_input_node
+                    const auto cardinality_in = _estimator->estimate_cardinality(disjunction_input_node);
+                    const auto cardinality_out = _estimator->estimate_cardinality(predicate_node);
+                    if (cardinality_out / cardinality_in > MAX_SELECTIVITY_FOR_PRE_JOIN_PREDICATE) return;
+                    predicate_node->set_left_input(nullptr);
+
+                    predicate_nodes.emplace_back(predicate_node);
+                  };
+
+              add_disjunction_if_beneficial(left_disjunction, join_node->left_input(), left_push_down_nodes);
+              add_disjunction_if_beneficial(right_disjunction, join_node->right_input(), right_push_down_nodes);
+
+              // End of the pre-join filter code
+            }
             _insert_nodes(current_node, input_side, {push_down_node});
-          }
-
-          if (move_to_left && move_to_right) {
+          } else if (move_to_left && move_to_right) {
             // Do not push down uncorrelated predicates
```

Comment:
"This predicate applies to both the left and the right side. We have not seen this case in the wild yet, it might make more sense to duplicate the predicate and push it down on both sides." ?

Reply:
Ack.